### ISSUE TITLE
Question about regex in Multiple Replace.

### ISSUE THREAD

**Issue by PopcornAWH** [SubtitleEdit/subtitleedit#3221]
Is there any documentation on how SE handles regex differently? When I try an  in SE working regex in an online regex tester, it doesn't catch any matches there.

Here's what I tried:
To add music symbols to make pairs on start & end of line, I made two rules:
Multi line add music symbols - find ^(♪ .+)(?:\n)(.*)(?<!$♪)$ replace $1 ♪\n♪ $2
Single line add music symbols - find ^(♪ .+)$(?<!♪)(?!\s) replace $1 ♪

When I check the multiline on regex101, it has no matches.

![capture](https://user-images.githubusercontent.com/45720608/49690359-cf044b80-fb2f-11e8-9614-4a144742a378.JPG)

**Comment by OmrSi** [comment]
SE doesn't support `multiline`.

**Comment by PopcornAWH** [comment]
I meant a subtitle with line break by multiline e.g.

♪ Come on, 
move your feet ♪

**Comment by mkver** [comment]
1. When I check your sample with regex101, it matches.
2. Your pattern looks strange: `(?<!$♪)` always returns true. You should remove `$ ` from it.
3. Even worse, your multiline pattern doesn't check whether there is already a ♪ at the end of the first or at the beginning of the second line. After applying the multiline regex, the regex still applies and SE therefore offers to add another pair of ♪. And what about cases like `♪ [Text] ♪\n [Text] ♪`?
4. The multiline option (and some other regex options) can be set inline (with `(?m)`). Or rather, it can be disabled with `(?-m)`, because multiline mode is [enabed by default](https://github.com/SubtitleEdit/subtitleedit/blob/master/src/Forms/MultipleReplace.cs#L233) in multiple replace.

   See [here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options) for more information about inline regex stuff.
5. Since #2716 SE only has one linebreak: `\n`. If you are using a normal regex engine, you have to differentiate between `\r`, `\n` and `\r\n`; not so in SE. At least that's how it should be.
6. But it seems that it was forgotten to also change $ in response to #2716. In
first line,
second line.
`^second` matches in multiple replace mode, but `line,$` does not, whereas `line,\r?$` does. My first try of fixing it was to replace a $ input by the user in a regex by `\r?$`. But this doesn't work in general, as $ is a zero-length match and `\r?$` is not (e.g. searching for `lin(e,\r?$)` gives a capture group that contains a linebreak). It is probably better to repace `$` by `(?=\r?$)`.
7. The default behaviour of the several regex searches differs: The find regex search does not use multiline mode by default, whereas both replace and multiple replace do.

**Comment by niksedk** [comment]
Source view + multiple replace uses multiline... other stuff does not - would it be easier if find/replace also used multiline? @mkver @OmrSi  ?

**Comment by OmrSi** [comment]
It would, but as you said, $ and \z are the same here because only $ is considered end of line.
So, it would only add the possibility for ^ to match the beginning of both lines in a line.
That's my understanding, at least. But yes, unifying them would help if someone is new to SE and doesn't know that there is a difference, or you can hint to it at the Regex section.

**Comment by PopcornAWH** [comment]
You guys go a bit fast for me, I'm still finding my way in regex. I am an avid SE user though, and always looking for tricks and tweaks to make my work easier. I think regex is a powerful tool, that's why I started learning how it works (and yes, still a long way to go).
I'll need to let your replies sink in. But thanks for taking the time to do so.

**Comment by PopcornAWH** [comment]
@mkver 
I took your comment to heart and tried to improve the regex for subtitle lines with a linebreak.

I think I managed that: 

https://regex101.com/r/S60aa3/1/  

it now doesn't offer more matches after applying. 
What I don't understand is, why this won't work if I take out the $ in the negative lookbehind as you suggested.

**Comment by mkver** [comment]
@OmrSi (and also @niksedk): What was said in #3165 is not really true: Multiline mode is enabled in some regex searches by default and where it is not, it can be activated inline. And the behaviour of `$` should be considered a bug because after #2716 SE is only supposed to treat all the different linebreaks the same. But it doesn't: `$` in multiline mode (like in multiple replace) matches to the left of an `\n`, but not to the left of `\r\n`. If you replace `$` by `(?=\r?$)` it should work. SE should do this automatically (at least on Windows, it should be unnecessary on systems where the default linebreak is simply `\n`), but currently doesn't.

And in multiline mode `$` and `\z` (or `\Z`) are not the same: `(?=\r?$)` matches before a newline or before the end of the string, but `(?=\r?$)` doesn't match in front of newlines in general. That you don't immediately see the difference is a bug, as has been said.
@niksedk:
1. I think it would be better to consistently use multiline everywhere. Someone who wants to match only the end should use `\z` or `\Z`.
(It might be nice to have beginning and end of line anchors as well as beginning and end of subtitle "packet" anchors in source view. I don't know if this is easily possible; if not, it might not be worth it. (I myself actually don't use source view at all and I don't know how many users do and how many of them use regular expressions at all.))
2. What does "source view uses multiline" actually mean? That the various regex searches use multiline mode when one is in source view? Because this is not consistent with my tests.
3. When I search in source view via ordinary search (Ctrl + F) or replace (Ctrl + H), there is a discrepancy between the results of the "Count" functionality that is part of search and the other search/replace functions. "Count" only searches the text (as if we were not in source view) and not the source. And it doesn't use multiline mode:
```
1
00:00:01,000 --> 00:00:03,000
First line,
second line;
third line.

2
00:00:03,080 --> 00:00:05,080
Fourth line
fifth line:
sixth line-


```
"Count" claims that there are two matches for `$` and `^` (so no multiline). Counting occurences of `^First` and `^Fourth` each have exactly one occurence according to "Count", and so have `line\.$` and `line-$`; whereas counting `00` gives zero occurences. So counting seems to ignore that I am in source view.

The behaviour of the ordinary (yet still regex) find and the replace function is different: It, too, doesn't seem to use multiline mode. But it doesn't find anything for the search for `^`; yet `$` finds a match. If the cursor is in front of the `1` at the beginning and I search for `(?s).*$` the message is that the match has been found at line number 0, yet the line with the 1 has line number 1. This is probably related to why `^` doesn't match. If I use the "Find" in search with this regex, then the initial `1` is not part of the match, but the linebreak after it is; if I use "Find" in the replace dialogue with the same regex (again when the cursor is in front of the initial `1`), then `1` is part of the match. This is very strange.

This is not restricted to the regex modes: If I search for `1` via "Find" in the search dialogue in normal mode, only the `1` that is part of `00:00:01,000` is found; with "Find" in the replace dialogue, it finds both matches.

**Comment by mkver** [comment]
@PopcornAWH: Your negative lookahead always matches, because `$♪` never matches. Leaving it out wouldn't change anything.

Leaving the `$` out is not the right approach either (sorry for that). The whole logic behind checking that the second line doesn't end in a `♪` is actuallly flawed: Your samples contain a `♪` at the end of the second line. In order to capture this format, you can use a positive lookbehind:
`^(♪ .+)(?<!♪)(?:\n)(.+)(?<=♪)$`
Notice that this regex presumes that there are no `\r` or that `\r\n` is treated as an indivisible unit by the regex engine that is matched by `\n` (otherwise your `.+` would also match the `\r` which implies that `(?<!♪)` matches automatically regardless of whether there is a `♪` at the position that a normal human would call "the end of the line").

Also notice that your new version still doesn't check whether there is a `♪` at the beginning of the second line.

**Comment by OmrSi** [comment]
@mkver As I am fairly new to this, this is my only issue, that multiline works in multiple replace but only for ^ and \A, and not for $ and \z (or \Z). Don't know if that a bug or something...

**Comment by mkver** [comment]
Multiline actually works in multipe replace as intended by the .NET regex engine that is behind this. What does not completely work is the extension by SE that aims to make `\n` a universal linebreak. And as already said: Use `(?=\r?$)` instead of `$` as a workaround.

**Comment by OmrSi** [comment]
@mkver I've been using `\n|$`, is it any different?

**Comment by OmrSi** [comment]
I meant (?=\n|$) instead of (?=\r?$), because I don't know how \r is handled.

> Maybe just replacing $ in the regex by (?=\r?$) is good enough?
If you mean by the user, then there must be a hint for it somewhere, as we assume that $ is end of line in both lines because ^ is start of line in both lines.

**Comment by mkver** [comment]
`(?=\n|$)` always matches before `\n`, whereas `(?=\r?$)` respects the current multiline-mode setting. I am actually thinking about automatically replacing (i.e. SE replaces, not the user) every `$` in the input with `(?=\r?$)`.

**Comment by aaaxx** [comment]
My eyes are bleeding.

Surely the only sane approach to newlines is to normalize them all to `\n` when opening files and pasting text, and to strip any `\r` automatically from all regexes.

Then you can optionally convert them to `Environment.NewLine` when saving. I say "optionally" because Microsoft has finally supported `\n` newlines in Notepad in [a recent Win 10 update](https://blogs.msdn.microsoft.com/commandline/2018/05/08/extended-eol-in-notepad/), so those people using it can at long last enjoy the benefits of a single intergalactic newline standard, as god intended. 😛

**Comment by OmrSi** [comment]
> (?=\n|$) always matches before \n, whereas (?=\r?$)

Oh, I see it now.

>  (i.e. SE replaces, not the user) every $ in the input with (?=\r?$).

Yeah, If that makes it like it's supposed to be, it would be great.

> My eyes are bleeding.

Imagine how I feel. 😂😂

**Comment by mkver** [comment]
One of the things I mentioned in #2716 was that singular `\r` or `\n` are not treated as linebreaks even in UTF-8. Maybe this is because of a limitation of the underlying .NET (on Windows)? If so, then aaaxx's suggestion is impossible, but it would still be possible to apply any regex to the string with any linebreak replaced by `\n` and converting it back after that.

**Comment by aaaxx** [comment]
I believe that's the result of the whole `Environment.NewLine` acrobatics used throughout the code: it defines linebreaks as `\r\n` when you run Subtitle Edit on Windows.

**Comment by PopcornAWH** [comment]
I really started something, causing all the bleeding eyes and worse ;-)

I noticed some srt files (the ones from the transcript datase) have the linux EOL (LF or \n I believe).
How would that affect the replacements in SE, or wouldn't it? 
![image](https://user-images.githubusercontent.com/45720608/49972343-9e5a5280-ff31-11e8-82c1-7033ab376a56.png)

**Comment by mkver** [comment]
When SE opens a file, it replaces the linebreaks with the ordinary linebreaks (i.e. `Enviroment.NewLine`).

**Comment by PopcornAWH** [comment]
Ah, you're right, should have checked that before posting, Open and save in SE does replace the EOL tags.

![image](https://user-images.githubusercontent.com/45720608/49973648-44a85700-ff36-11e8-92fa-d33ca70327a0.png)

**Comment by OmrSi** [comment]
@PopcornAWH As I said, I am fairly new to this, so I wrote these [Music Test.zip](https://github.com/SubtitleEdit/subtitleedit/files/2678176/Music.Test.zip)
There are two for each case to include both cases of space or no space, I don't know if there is a way to merge them.

I tested them on the lines you attached, they worked for me.
@mkver I hope you can take a look and tell if you have any advice about them and if there is a way to merge them so I can improve myself.

**Comment by PopcornAWH** [comment]
@OmrSi 

Here's what it did with my test sub. 

Add to beginning, no space -> Match line 2
Add to beginning, space -> No match
Add to end, no space -> Match lines 1,2,4,7,8,9,10,11
Add to end, space -> No  match

It doesn't catch line 17, but that can be easily done by removing the caret, I do this only with caution using the preview.
![music test-1](https://user-images.githubusercontent.com/45720608/49993882-35ea9000-ff88-11e8-924d-7d52a2818f6a.jpg)
![music test-2](https://user-images.githubusercontent.com/45720608/49993899-43a01580-ff88-11e8-8696-7d0b24ecde5d.jpg)
![music test-3](https://user-images.githubusercontent.com/45720608/49993956-721df080-ff88-11e8-8e45-feedc5f6a3c0.jpg)

I think the rule Add to end, space will never match, as SE removes a space at the end of the line (does for me anyway)

**Comment by OmrSi** [comment]
> space will never match, as SE removes a space at the end of the line

It does when you break a line after a space and don't delete it manually. That case is only fixed when using Fix common error. and in that case, the space at the beginning would be deleted too.
So it has its benefits if you don't use Fix common errors, but if you do, then both cases of space at the beginning and at the end are useless.

As for case 11 and 17, I didn't know those existed from your original examples.
I thought it has to be at the beginning, but in that case, you can delete the carets from both expressions.

Then you'd have to cases left, first one is adding a symbol before the [], as per No. 11.
Second is adding if after a speaker's name if there is one at the end, I don't know if you're interested in that or just adding to the very beginning or end is enough for you. :D

**Comment by OmrSi** [comment]
Yeah. it's not easy. :\

**Comment by OmrSi** [comment]
After working more on it: [Music Test 3.zip](https://github.com/SubtitleEdit/subtitleedit/files/2683202/Music.Test.3.zip)

**Comment by PopcornAWH** [comment]
@OmrSi 
Test 2 works fine! Great work, will take a look at Test3 now,

@mkver 

I used ^[A-Z ]+$ to add brackets to the HI in english shows. (they don't do brackets sometimes).
This works good, as the HI never has punctuation. But with your alternative for $ it works even better.

![find hi1](https://user-images.githubusercontent.com/45720608/50052330-c1982400-0122-11e9-9ce9-645adc705ec2.jpg)

![find hi2](https://user-images.githubusercontent.com/45720608/50052331-c65cd800-0122-11e9-946e-15546813bab1.jpg)

It also catches the first line on a subtitle line with a linebreak. (that never occurs, but anyway, it's more accurate).

Thanks guys for your time and effort here.

**Comment by OmrSi** [comment]
Test 3 includes dialog, HI and speaker. 
For example,  a line like this:

- WALKER: [SIGHS] How are you? ♪
- ANNE: Good. ♪

Would become: 
- WALKER: [SIGHS]♪How are you? ♪
- ANNE:♪ Good. ♪


It took me a while to make the Regex scan the whole line for a second ♪, not only at the beginning and end, but I think the results are satisfying. 
In the end, it depends on you file, how the HI text is written and all that…

**Comment by OmrSi** [comment]
@PopcornAWH Improved it to include more than one word inside []: [Music Test 4.zip](https://github.com/SubtitleEdit/subtitleedit/files/2683549/Music.Test.4.zip)

**Comment by OmrSi** [comment]
I'm trying to use this in SE: https://regex101.com/r/31Ne1c/2
but it's not working. :\
It seems that The Arabic comma is causing a problem.
![screenshot 804](https://user-images.githubusercontent.com/20923700/50095147-79116100-021d-11e9-837d-ac26d751c7aa.png)

**Comment by PopcornAWH** [comment]
@OmrSi 
I tried (\d+)\s([،])\s?(\d+)  and that worked.

**Comment by OmrSi** [comment]
It isn't working here, see the other space is still there: 
![screenshot 808](https://user-images.githubusercontent.com/20923700/50101785-25f2da80-022c-11e9-97a2-87fc3664c5f9.png)

**Comment by mkver** [comment]
The Arabic comma seems to change the appearance of the closing bracket after it (makes it look like opening bracket -- probably because Arabic is written right-to-left, but I can't say whether this is a bug or whether it is intended behaviour), but apart from this the regex worked. I don't have a space after the comma.

Have you checked whether another active regex is interfering? And could you please upload your settings.xml?

**Comment by OmrSi** [comment]
@mkver Thank you! Turns out another one is indeed interfering, I turned the rest off and it worked, silly me! 😅
Sorry for the trouble.

But yes, Arabic comma always changes the appearance of every kind of bracket, here and in the text box, though it doesn't in other programs.

**Comment by OmrSi** [comment]
@PopcornAWH Hey, I know it's been a while.
I've been learning some regex in the last couple of months and was able to reduce all of the regexes I wrote to 3 lines!
Thought I'd share it with ya: [Fix music notes.zip](https://github.com/SubtitleEdit/subtitleedit/files/3050960/Fix.music.notes.zip)
